R/install_torsten.R

```diff
@@ -0,0 +1,197 @@
+  if (release_list_downloaded != 0) {
+    stop("GitHub download of release list failed.", call. = FALSE)
+  }
+  release <- jsonlite::read_json(dest_file)
+  sapply(release,
+         function(x){
+           x$tag_name
+         })
+}
+
+install_torsten <- function(dir = NULL,
+                            cores = getOption("mc.cores", 2),
+                            quiet = FALSE,
+                            timeout = 1200,
+                            version = NULL,
+                            release_url = NULL,
+                            cpp_options = list(),
+                            check_toolchain = TRUE
+) {
+
+  ## install_torsten is based on cmdstanr::install_cmdstan. Differences include
+  ## * install_torsten always overwrites pre-existing Torsten installation at the
+  ##.  specified cmdstan_path.
+  ## * install_torsten does not try to automatically append to cpp_options for
+  ##.  M1 Macs. Is this needed?
+  ## * install_torsten does not provide support for WSL. Should it?
+
+  require(cmdstanr)
+  if (check_toolchain) {
+    cmdstanr::check_cmdstan_toolchain(fix = FALSE, quiet = quiet)
+  }
+  if (is.null(dir)) {
+    dir <- file.path(dirname(cmdstanr::cmdstan_default_install_path()),
+                     ".torsten")
+  }
+  if (!dir.exists(dir)) {
+    dir.create(dir, recursive = TRUE)
+  }
+  if (!is.null(version)) {
+    if (!is.null(release_url)) {
+      warning("version and release_url shouldn't both be specified!",
+              "\nrelease_url will be ignored.", call. = FALSE)
+    }
+
+    release_list <- get_torsten_release_list()
+    release <- release_list[grep(version, release_list)]
+    if(length(release) < 1){
+      stop("Available Torsten versions do not include ", version, call. = FALSE)
+    }
+    download_url <- paste0("https://github.com/metrumresearchgroup/Torsten/archive/refs/tags/",
+                          release, ".tar.gz")
+  }
+  if (!is.null(release_url)) {
+    if (!endsWith(release_url, ".tar.gz")) {
+      stop(release_url, " is not a .tar.gz archive!",
+           "cmdstanr supports installing from .tar.gz archives only.", call. = FALSE)
+    }
+    download_url <- release_url
+    message("* Installing Torsten from ", release_url)
+  } else {
+    if(is.null(version)) {
+      ver <- latest_torsten_release()
+      message("* Latest Torsten release is ", ver)
+      download_url <- paste0("https://github.com/metrumresearchgroup/Torsten/archive/refs/tags/",
+                            ver, ".tar.gz")
+    }
+  }
+  message("* Installing Torsten from ", download_url)
+  dir_torsten <- file.path(dir, substr(basename(download_url), 1, nchar(basename(download_url)) - 7))
+  dir.create(dir_torsten, recursive = TRUE, showWarnings = FALSE)
+  dest_file <- file.path(dir_torsten, basename(download_url))
+  dir_cmdstan <- file.path(dir_torsten, "cmdstan")
+  ## Reset timeout for download. The 60 s default is not enough.
+  options(timeout = max(300, getOption("timeout")))
```

Comment:
This should restore `timeout` to its original value (e.g., using [withr::with\_options](https://withr.r-lib.org/reference/with_options.html)).

Reply:
Done